master/getting-started/kubernetes/installation/hardway/install-cni-plugin.md

```diff
@@ -0,0 +1,184 @@
+
+The CNI Plugin interacts with the Kubernetes APIServer while creating pods, both to obtain additional information
+and to update the data store with information about the pod.
+
+On the Kubernetes master node, create a key for the CNI Plugin to authenticate with and certificate signing request.
+
+```
+openssl req -newkey rsa:4096 \
+           -keyout cni.key \
+           -nodes \
+           -out cni.csr \
+           -subj "/CN=calico-cni"
+```
+
+We will sign this certificate using the main Kubernetes CA.
+
+```
+sudo openssl x509 -req -in cni.csr \
+                  -CA /etc/kubernetes/pki/ca.crt \
+                  -CAkey /etc/kubernetes/pki/ca.key \
+                  -CAcreateserial \
+                  -out cni.crt \
+                  -days 365
+sudo chown ubuntu:ubuntu cni.crt
+```
+
+Next, we create a kubeconfig file for the CNI.
+
+```
+APISERVER=$(kubectl config view -o jsonpath='{.clusters[0].cluster.server}')
+kubectl config set-cluster kubernetes \
+    --certificate-authority=/etc/kubernetes/pki/ca.crt \
+    --embed-certs=true \
+    --server=$APISERVER \
+    --kubeconfig=cni.kubeconfig
+
+kubectl config set-credentials calico-cni \
+    --client-certificate=cni.crt \
+    --client-key=cni.key \
+    --embed-certs=true \
+    --kubeconfig=cni.kubeconfig
+
+kubectl config set-context default \
+    --cluster=kubernetes \
+    --user=calico-cni \
+    --kubeconfig=cni.kubeconfig
+
+kubectl config use-context default --kubeconfig=cni.kubeconfig
+```
+
+Copy this `cni.kubeconfig` file to every node in the cluster.
+
+## Provision RBAC
+
+Define a ClusterRole the CNI Plugin will use to access Kubernetes.
+
+```
+kubectl apply -f - <<EOF 
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: calico-cni
+rules:
+  # The CNI plugin needs to get pods, nodes, and namespaces.
+  - apiGroups: [""]
+    resources:
+      - pods
+      - nodes
+      - namespaces
+    verbs:
+      - get
+  # The CNI plugin patches pods/status.
+  - apiGroups: [""]
+    resources:
+      - pods/status
+    verbs:
+      - patch
+ # These permissions are required for Calico CNI to perform IPAM allocations.
+  - apiGroups: ["crd.projectcalico.org"]
+    resources:
+      - blockaffinities
+      - ipamblocks
+      - ipamhandles
+    verbs:
+      - get
+      - list
+      - create
+      - update
+      - delete
+  - apiGroups: ["crd.projectcalico.org"]
+    resources:
+      - ipamconfigs
+      - clusterinformations
+      - ippools
+    verbs:
+      - get
+      - list
+EOF
+```
+
+Bind the ClusterRole to the `calico-cni` account.
+
+```
+kubectl create clusterrolebinding calico-cni --clusterrole=calico-cni --user=calico-cni
+```
+
+## Install the plug in
+
+Do these steps on each node in your cluster.
+
+Run these commands as root.
+
+```
+sudo su
+```
+
+Install the CNI Plugin Binaries
+
+```
+curl -L -o /opt/cni/bin/calico https://github.com/projectcalico/cni-plugin/releases/download/v3.8.0/calico-amd64
```

Comment:
We should use the automatic URLs for these as well so they aren't hard coded to a particular version and instead use the page version.

Reply:
Right, we don't have master builds of this.

IIRC elsewhere we've pulled the docker image and copied the binaries to the host from that (we do have master builds of the CNI plugin docker image)